### PR TITLE
Add support for dynamic widgets in Forms

### LINKED ISSUE #2236: Add initial support for EEF's DynamicFor/If to Forms
Sirius Desktop, via EEF, supports [dynamic mappings](https://eclipse.dev/sirius/doc/specifier/properties/Properties_View_Description.html#dynamic_mappings) that offer the specifier/studio maker powerful control on how many instances of a given widget (including zero) are instanciated when a Form is rendered.

We will port this capability to the Sirius Component's Forms, including the View DSL.
As a first step (this ticket), we will implement the exact same feature as is present in EEF.
We'll see afterwards if/how it can be generalized.

### PR DESCRIPTION
**Pull request by pcdavid**

### CODE REVIEW

File: packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/ButtonComponent.java [hunk 1]
```diff
@@ -45,8 +45,15 @@ public Element render() {
         VariableManager variableManager = this.props.getVariableManager();
         ButtonDescription buttonDescription = this.props.getButtonDescription();
 
-        String id = buttonDescription.getIdProvider().apply(variableManager);
         String label = buttonDescription.getLabelProvider().apply(variableManager);
+
+        VariableManager idVariableManager = variableManager.createChild();
+        idVariableManager.put(FormComponent.TARGET_OBJECT_ID, buttonDescription.getTargetObjectIdProvider().apply(variableManager));
+        idVariableManager.put(FormComponent.TARGET_OBJECT_ID, buttonDescription.getTargetObjectIdProvider().apply(variableManager));
```
Comment (frouene) [inline packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/ButtonComponent.java]: Duplicate lines

File: packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/FormComponent.java [hunk 2]
```diff
@@ -31,10 +31,28 @@
...
+     * The name of the variable used to store the id of widget's description in the {@link VariableManager} for a widget
+     * to compute its own id.
+     */
+    public static final String TARGET_OBJECT_ID = "targetObjectId";
+
+    /**
+     * The name of the variable used to store the id of widget's description in the {@link VariableManager} for a widget
+     * to compute its own id.
+     */
+    public static final String WIDGET_LABEL = "widgetLabel";
```
Comment (frouene) [inline packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/FormComponent.java]: It's the same comment for the three variables

File: packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/FormComponent.java [hunk 3]
```diff
@@ -65,23 +88,15 @@ public Element render() {
...
-                .toList();
+        List<Element> children = candidates.stream().map(candidate -> {
+            VariableManager childVariableManager = variableManager.createChild();
+            childVariableManager.put(VariableManager.SELF, candidate);
+            childVariableManager.put(FormComponent.PARENT_ELEMENT_ID, id);
+            return childVariableManager;
+        }).flatMap(childVariableManager -> pageDescriptions.stream().filter(pageDescription -> pageDescription.getCanCreatePredicate().test(childVariableManager)).map(pageDescription -> {
+            PageComponentProps pageComponentProps = new PageComponentProps(childVariableManager, pageDescription, this.props.getWidgetDescriptors());
+            return new Element(PageComponent.class, pageComponentProps);
+        })).toList();
```
Comment (frouene) [inline packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/FormComponent.java]: Indentation doesn't seem right

File: packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/description/IfDescription.java [hunk 4]
```diff
@@ -25,27 +26,22 @@
...
  */
 @Immutable
-public final class IfDescription {
-    private String id;
+public final class IfDescription extends AbstractControlDescription {
 
     private Function<VariableManager, Boolean> predicate;
 
-    private AbstractWidgetDescription widgetDescription;
+    private List<AbstractControlDescription> childrenDescription;
```
Comment (frouene) [inline packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/description/IfDescription.java]: In ForDescription, it's named controlDescriptions

File: packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/diagram/NodeStylePropertiesConfigurer.java [hunk 5]
```diff
@@ -89,13 +89,16 @@ public class NodeStylePropertiesConfigurer implements IPropertiesDescriptionRegi
...
+    private final Function<VariableManager, String> semanticTargetIdProvider;
+
     public NodeStylePropertiesConfigurer(ICustomImageMetadataSearchService customImageSearchService, IValidationService validationService,
             List<IParametricSVGImageRegistry> parametricSVGImageRegistries, AQLTextfieldCustomizer aqlTextfieldCustomizer, IObjectService objectService) {
         this.validationService = Objects.requireNonNull(validationService);
         this.customImageSearchService = Objects.requireNonNull(customImageSearchService);
         this.parametricSVGImageRegistries = parametricSVGImageRegistries;
         this.aqlTextfieldCustomizer = Objects.requireNonNull(aqlTextfieldCustomizer);
         this.objectService = Objects.requireNonNull(objectService);
+        this.semanticTargetIdProvider = variableManager -> variableManager.get(VariableManager.SELF, Object.class).map(objectService::getId).orElse(null);
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/diagram/NodeStylePropertiesConfigurer.java:101]: I don't have a strong opinion on the matter, but for semanticElementsProvider we declared its value directly at initialization and not in the constructor.

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/impl/FormElementDescriptionImpl.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2023 Obeo.
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/impl/FormElementDescriptionImpl.java]: New file, should be only 2023

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/impl/FormElementForImpl.java [hunk 7]
```diff
@@ -0,0 +1,285 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2023 Obeo.
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/impl/FormElementForImpl.java]: New file, should be only 2023

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/impl/FormElementIfImpl.java [hunk 8]
```diff
@@ -0,0 +1,229 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2023 Obeo.
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/impl/FormElementIfImpl.java]: New file, should be only 2023

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/ControlDescription.java [hunk 9]
```diff
@@ -0,0 +1,55 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2023 Obeo.
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/ControlDescription.java]: New file, should be only 2023

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/FormElementDescription.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2023 Obeo.
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/FormElementDescription.java]: New file, should be only 2023

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/FormElementFor.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2023 Obeo.
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/FormElementFor.java]: New file, should be only 2023

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/FormElementIf.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2023 Obeo.
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/FormElementIf.java]: New file, should be only 2023

File: packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormElementDescriptionItemProvider.java [hunk 13]
```diff
@@ -0,0 +1,140 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2023 Obeo.
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormElementDescriptionItemProvider.java]: New file, should be only 2023

File: packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormElementForItemProvider.java [hunk 14]
```diff
@@ -0,0 +1,214 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2023 Obeo.
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormElementForItemProvider.java]: New file, should be only 2023

File: packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormElementIfItemProvider.java [hunk 15]
```diff
@@ -0,0 +1,201 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2023 Obeo.
```
Comment (frouene) [inline packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormElementIfItemProvider.java]: New file, should be only 2023

File: doc/adrs/107_dynamic_widgets.adoc [hunk 16]
```diff
@@ -11,29 +11,32 @@ The core rendering algorithm of the Sirius Web Forms already implements these me
 
 == Decision
 
-We will add support for `DynamicFor` and `DynamicIf` to the Forms DSL.
+We will add support for _For_ and _If_ elements to the Forms DSL.
+They will be named _FormElementFor_ and _FormElementIf_ in the metamodel, but presented as simply _For_ an _If_ in the UI.
+In this first version, we will only support these new constructs inside _Groups_ and _FlexboxContainers_.
 
-In this first version, we will only support the same case as in EEF desktop, namely widgets inside a _Group_.
+All these four elements (For, If, Groups, Flexbox) behave as containers for an arbitrary number sub-elements.
```
Comment (frouene) [inline doc/adrs/107_dynamic_widgets.adoc]: number of

File: doc/adrs/107_dynamic_widgets.adoc [hunk 17]
```diff
@@ -11,29 +11,32 @@ The core rendering algorithm of the Sirius Web Forms already implements these me
...
-
-_DynamicFor_ will have two attributes:
+_FormElementFor_ will have two attributes:
 
 * `iterationExpression`: an AQL expression to be evaluated when rendering the form, which should return a collection of (arbitrary) elements;
 * `iterator`: a (fixed) variable name.
 
-It will also _contain_ an arbitrary number if _DynamicIf_, which has a `predicateExpression` attribute and a `control: ControlDescription[1]` containment reference.
+The semantics of the element is that it's children element will be rendered once for every element in the collection returned by `iterationExpression`, in order.
+For each value _V_ returned by `iterationExpression`, all the children are rendered in a context where _V_ is bound to a variable named as specified buy _iterator_.
```
Comment (frouene) [inline doc/adrs/107_dynamic_widgets.adoc]: buy -> by

File: CHANGELOG.adoc [hunk 18]
```diff
@@ -29,6 +29,12 @@ All of its remaining content was only relevant inside of Sirius Web and it has t
...
 - https://github.com/eclipse-sirius/sirius-web/issues/2296[#2296] [diagram] `sirius-components-diagrams-reactflow` now depends on `html-to-image` node package.
+- https://github.com/eclipse-sirius/sirius-web/issues/2236[#2236] [forms] The introduction of _ForDescription_ and _IfDescription_ elements to the Forms DSL introduced breaking changes in the metamodel.
+Existing FormDescription documents will need to be modified to load correctly, and any custom widget's metamodel will need to be regenerated using the new `form.ecore` version as reference.
+See [ADR-107] _Add support for dynamic and conditional widgets in Forms_ for the details of the changes.
+Also, all widget descriptions now have a `Function<VariableManager, String> targetObjectIdProvider`, needed to compute the widget instance's id.
+The new `targetObjectIdProvider` is required to be non-null, so any code which programatically builds core Form Descriptions (not View-based ones), needs to pass a non-null provider.
```
Comment (frouene) [inline CHANGELOG.adoc]: programatically -> programmatically

**Review by frouene — CHANGES_REQUESTED** [review]

File: packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/ButtonComponent.java (hunk 1 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/ButtonComponent.java]: Done.

File: packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/FormComponent.java (hunk 2 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/FormComponent.java]: Done.

File: packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/FormComponent.java (hunk 3 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/components/FormComponent.java]: Done.

File: packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/description/IfDescription.java (hunk 4 above)
Comment (pcdavid) [reply packages/forms/backend/sirius-components-forms/src/main/java/org/eclipse/sirius/components/forms/description/IfDescription.java]: Done.

File: packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/diagram/NodeStylePropertiesConfigurer.java (hunk 5 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-emf/src/main/java/org/eclipse/sirius/components/view/emf/diagram/NodeStylePropertiesConfigurer.java:101]: We can't in this case as we need the `objectService` which is only initialized in the constructor.

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/impl/FormElementDescriptionImpl.java (hunk 6 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/impl/FormElementDescriptionImpl.java]: Done.

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/impl/FormElementForImpl.java (hunk 7 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/impl/FormElementForImpl.java]: Done.

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/impl/FormElementIfImpl.java (hunk 8 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/impl/FormElementIfImpl.java]: Done.

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/ControlDescription.java (hunk 9 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/ControlDescription.java]: Done.

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/FormElementDescription.java (hunk 10 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/FormElementDescription.java]: Done.

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/FormElementFor.java (hunk 11 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/FormElementFor.java]: Done.

File: packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/FormElementIf.java (hunk 12 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-form/src/main/java/org/eclipse/sirius/components/view/form/FormElementIf.java]: Done.

File: packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormElementDescriptionItemProvider.java (hunk 13 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormElementDescriptionItemProvider.java]: Done.

File: packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormElementForItemProvider.java (hunk 14 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormElementForItemProvider.java]: Done.

File: packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormElementIfItemProvider.java (hunk 15 above)
Comment (pcdavid) [reply packages/view/backend/sirius-components-view-form-edit/src/main/java/org/eclipse/sirius/components/view/form/provider/FormElementIfItemProvider.java]: Done.

File: doc/adrs/107_dynamic_widgets.adoc (hunk 16 above)
Comment (pcdavid) [reply doc/adrs/107_dynamic_widgets.adoc]: Done.

File: doc/adrs/107_dynamic_widgets.adoc (hunk 17 above)
Comment (pcdavid) [reply doc/adrs/107_dynamic_widgets.adoc]: Done.

File: CHANGELOG.adoc (hunk 18 above)
Comment (pcdavid) [reply CHANGELOG.adoc]: Done.

**Review by frouene — APPROVED** [review]

**Pull request merged**